HW3.py

```diff
@@ -0,0 +1,253 @@
+
+
+print("""#1
+Option 1
+(!) This solution was made before any hints""")
+print("Here it was: ", random_line)
+print("Here it became: ", consonants_to_vowels_1(random_line))
+
+
+consonants = "bcdfghjklmnpqrstvwxyz"
+vowels = "aeiou"
+
+def consonants_to_vowels_2(the_string):
+    import random
+    for the_consonant in consonants:
+        # print(the_string)
+        the_string = the_string.lower().replace(the_consonant, random.choice(vowels), random_line.count(the_consonant))
+    return the_string
+
+print("""#1
+Option 2
+The option 2 was made after I heard about the find(), so I decided to use it
+I still think that the first option is better because it "picks up" the string less times""")
+print("Here it was: ", random_line)
+print("Here it became: ", consonants_to_vowels_2(random_line))
+
+
+
+
+# 2
+# 2) Дан массив из словарей
+# data = [
+#   {'name': 'Viktor', 'city': 'Kiev', 'age': 30 },
+#   {'name': 'Maksim', 'city': 'Dnepr', 'age': 20},
+#   {'name': 'Vladimir', 'city': 'Lviv', 'age': 32},
+#   {'name': 'Andrey', 'city': 'Kiev', 'age': 34},
+#   {'name': 'Artem', 'city': 'Dnepr', 'age': 50},
+#   {'name': 'Dmitriy', 'city': 'Lviv', 'age': 21}]
+#
+# 2.1) отсортировать массив из словарей по значению ключа ‘age'
+# # use sorted()
+# 2.2) сгруппировать данные по значению ключа 'city'
+# вывод должен быть такого вида :
+# result = {
+#   'Kiev': [
+#    {'name': 'Viktor', 'age': 30 },
+#    {'name': 'Andrey', 'age': 34}],
+#
+#   'Dnepr': [ {'name': 'Maksim', 'age': 20 },
+#        {'name': 'Artem', 'age': 50}],
+#   'Lviv': [ {'name': 'Vladimir', 'age': 32 },
+#        {'name': 'Dmitriy', 'age': 21}]
+# }
+
+data = [
+  {'name': 'Viktor', 'city': 'Kiev', 'age': 30 },
+  {'name': 'Maksim', 'city': 'Dnepr', 'age': 20},
+  {'name': 'Vladimir', 'city': 'Lviv', 'age': 32},
+  {'name': 'Andrey', 'city': 'Kiev', 'age': 34},
+  {'name': 'Artem', 'city': 'Dnepr', 'age': 50},
+  {'name': 'Dmitriy', 'city': 'Lviv', 'age': 21}]
+# 2.1) отсортировать массив из словарей по значению ключа ‘age'
+def sort_by_age(list_of_dicts):
+    return sorted(list_of_dicts, key=lambda dict_elem: dict_elem['age'])
+print(f""" #2.1
+I'm still not big fan of lambdas, probably I need to rewrite it using an ordinary function
+BTW there is no honor in this solution because I took it from https://tproger.ru/translations/python-sorting/
+2.1 answer is:
+{sort_by_age(data)}""")
+
+# 2.2) сгруппировать данные по значению ключа 'city'
+# вывод должен быть такого вида :
+# result = {
+#   'Kiev': [
+#    {'name': 'Viktor', 'age': 30 },
+#    {'name': 'Andrey', 'age': 34}],
+
+#   'Dnepr': [ {'name': 'Maksim', 'age': 20 },
+#        {'name': 'Artem', 'age': 50}],
+#   'Lviv': [ {'name': 'Vladimir', 'age': 32 },
+#        {'name': 'Dmitriy', 'age': 21}]
+# }
+
+print("""
+#2.2 is WRONG
+I don't know why `result[element['city']].append({'name':element['name'], 'age':element['age']})`
+appends to ALL THREE cities, not only to result[element['city']]
```

Comment:
if you try to print the result out of `for` then you will see that not 3, but 6 elements are added to each city. That is because you iterate over data_copy which has 6 elements, which means 6 times to add. You need to add some `if` and check. 
<img width="621" alt="Снимок экрана 2022-03-31 в 22 22 52" src="https://user-images.githubusercontent.com/44474670/161133351-02c79039-0598-4c25-9ee9-682d2a227f38.png">
```
def group_by_city(some_arg):
    result_data = {}
    for item in some_arg:
        city = item.pop('city')
        if city in result_data:
            result_data[city].append(item)
        else:
            result_data[city] = [item]
    return result_data
print(group_by_city(data))
```

Reply:
Thanks for the explanation; Sorry, I didn't notice your code until now (after the #2.2 solution comminment) so did a very similar thing by myself.